Fixes #79187 Add MemoryMappedFile.CreateFromFile(SafeFileHandle) overload

amalabey opened the pull request:
Fixes #79187

amalabey committed:
Add CreateFromFile overload with SafeFileHandle

The overload uses the handle the perform validations instead of
instantiating a FileStream and calling the FileStream overload,
because the FileStream instantiation appears to be heavy on allocations.

amalabey committed:
Add invalid capacity tests

amalabey committed:
Refactor test name to align with existing

amalabey committed:
Increase test coverage for SafeFileHandle

amalabey committed:
Merge branch 'dotnet:main' into issue-79187

amalabey committed:
Add API docs for the new overload

amalabey committed:
Remove unnecessary line

amalabey commented:
@dotnet-policy-service agree

adamsitnik commented on src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.cs:
```diff
@@ -186,6 +186,92 @@ public static MemoryMappedFile CreateFromFile(string path, FileMode mode, string
             return new MemoryMappedFile(handle, fileHandle, false);
         }
 
+        /// <summary>
+        /// Creates a memory-mapped file from an existing file using a SafeFileHandle,
+        /// and the specified access mode, name, inheritability, and capacity.
+        /// </summary>
+        /// <param name="fileHandle">The SafeFileHandle to the existing file. Caller is
+        /// responsible for disposing <c>fileHandle</c> when <paramref name="leaveOpen"/>==true (otherwise,
+        /// automatically disposed by the <c>MemoryMappedFile</c>). </param>
+        /// <param name="mapName">A name to assign to the memory-mapped file, or <c>null</c> for a
+        /// <see cref="MemoryMappedFile"/> that you do not intend to share across processes.</param>
+        /// <param name="capacity">The maximum size, in bytes, to allocate to the memory-mapped file.
+        /// Specify 0 to set the capacity to the size of <c>filestream</c>.</param>
+        /// <param name="access">One of the enumeration values that specifies the type of access allowed
+        /// to the memory-mapped file.<para>This parameter can't be set to <c>Write</c>.</para></param>
+        /// <param name="inheritability">One of the enumeration values that specifies whether a handle
+        /// to the memory-mapped file can be inherited by a child process. The default is <c>None</c>.</param>
+        /// <param name="leaveOpen">A value that indicates whether to close the source file handle when
+        /// the <see cref="MemoryMappedFile"/> is disposed.</param>
+        /// <returns>A memory-mapped file that has the specified characteristics.</returns>
+        /// <exception cref="ArgumentException">
+        /// <para><c>mapName</c> is <c>null</c> or an empty string.</para>
+        /// <para>-or-
+        /// <c>capacity</c> and the length of the file are zero.</para>
+        /// <para>-or-
+        /// <c>access</c> is set to Write or Write enumeration value, which is not allowed.</para>
+        /// <para>-or-
+        /// access is set to Read and <c>capacity</c> is larger than the length of <c>fileHandle</c>.</para>
+        /// </exception>
+        /// <exception cref="ArgumentNullException"><c>fileHanlde</c> is <c>null</c>.</exception>
+        /// <exception cref="ArgumentOutOfRangeException">
+        /// <c>capacity</c> is less than zero.
+        /// <para>-or-
+        /// <c>capacity</c> is less than the file size.</para>
+        /// <para>-or-
+        /// <c>access</c> is not a valid <see cref="MemoryMappedFileAccess"/> enumeration value.</para>
+        /// <para>-or-
+        /// <c>inheritability</c> is not a valid <see cref="HandleInheritability"/> enumeration value.</para>
+        /// </exception>
```
The xml docs need some minor improvements:
* instead of `==true` we need to use `<see langword="true" />`
* instead of `<c>null</c>` we need to use `<see langword="null" />`
* when refering to parameters we need to use `<paramref name="$name" />` syntax

amalabey replied:
Fixed

adamsitnik commented on src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.cs:
```diff
@@ -186,6 +186,92 @@ public static MemoryMappedFile CreateFromFile(string path, FileMode mode, string
+        {
+            ArgumentNullException.ThrowIfNull(fileHandle);
+
+            if (mapName != null && mapName.Length == 0)
+            {
+                throw new ArgumentException(SR.Argument_MapNameEmptyString);
+            }
+
+            if (capacity < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(capacity), SR.ArgumentOutOfRange_PositiveOrDefaultCapacityRequired);
+            }
+
+            long fileSize = RandomAccess.GetLength(fileHandle);
+            if (capacity == 0 && fileSize == 0)
+            {
+                throw new ArgumentException(SR.Argument_EmptyFile);
+            }
+
+            if (access < MemoryMappedFileAccess.ReadWrite ||
+                access > MemoryMappedFileAccess.ReadWriteExecute)
+            {
+                throw new ArgumentOutOfRangeException(nameof(access));
+            }
+
+            if (access == MemoryMappedFileAccess.Write)
+            {
+                throw new ArgumentException(SR.Argument_NewMMFWriteAccessNotAllowed, nameof(access));
+            }
+
+            if (inheritability < HandleInheritability.None || inheritability > HandleInheritability.Inheritable)
+            {
+                throw new ArgumentOutOfRangeException(nameof(inheritability));
+            }
```
the  changes you are introduced are perfectly correct. Would you be interested in moving the argument validation logic to a separate method and reusing it in all 3 the `CreateFromFile` overloads (accepting path, FileStream and SafeFileHandle)?

amalabey replied:
Absolutely, that makes sense. I will make that refactoring and update the PR.

amalabey replied:
Addressed

adamsitnik reviewed (commented):
The changes look very good to me, thank you for you contribution @amalabey!

I need two things before I hit the approve button:
* minor improvements of the xml doc comments (please see my comment)
* the conflict needs to be solved

amalabey committed:
Merge remote-tracking branch 'upstream/main' into issue-79187

amalabey committed:
Fix xml doc issues

amalabey committed:
Fix paramref and param mistakes

amalabey committed:
Separate and reuse validation logic

amalabey committed:
Tag failing test with the related active issue

amalabey committed:
Merge remote-tracking branch 'upstream/main' into issue-79187

amalabey commented:
@adamsitnik Addressed your feedback. I had to mark a test method with ActiveIssueAttribute, because the related CreateFromFile(FileStream) test was failing due to the same reason.

adamsitnik commented on src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.cs:
```diff
@@ -186,37 +167,83 @@ public static MemoryMappedFile CreateFromFile(string path, FileMode mode, string
         }
 
-        public static MemoryMappedFile CreateFromFile(FileStream fileStream, string? mapName, long capacity,
+        /// <summary>
+        /// Creates a memory-mapped file from an existing file using a SafeFileHandle,
+        /// and the specified access mode, name, inheritability, and capacity.
+        /// </summary>
+        /// <param name="fileHandle">The SafeFileHandle to the existing file. Caller is
+        /// responsible for disposing <paramref name="fileHandle"/> when <paramref name="leaveOpen"/> is <see langword="true" /> (otherwise,
+        /// automatically disposed by the <see cref="MemoryMappedFile"/>). </param>
+        /// <param name="mapName">A name to assign to the memory-mapped file, or <see langword="null" /> for a
+        /// <see cref="MemoryMappedFile"/> that you do not intend to share across processes.</param>
+        /// <param name="capacity">The maximum size, in bytes, to allocate to the memory-mapped file.
+        /// Specify 0 to set the capacity to the size of the file.</param>
+        /// <param name="access">One of the enumeration values that specifies the type of access allowed
+        /// to the memory-mapped file.<para>This parameter can't be set to <c>Write</c>.</para></param>
+        /// <param name="inheritability">One of the enumeration values that specifies whether a handle
+        /// to the memory-mapped file can be inherited by a child process. The default is <c>None</c>.</param>
+        /// <param name="leaveOpen">A value that indicates whether to close the source file handle when
+        /// the <see cref="MemoryMappedFile"/> is disposed.</param>
+        /// <returns>A memory-mapped file that has the specified characteristics.</returns>
+        /// <exception cref="ArgumentException">
+        /// <para><paramref name="mapName"/> is <see langword="null" /> or an empty string.</para>
+        /// <para>-or-
+        /// <paramref name="capacity"/> and the length of the file are zero.</para>
+        /// <para>-or-
+        /// <paramref name="capacity"/> is set to Write or Write enumeration value, which is not allowed.</para>
+        /// <para>-or-
+        /// access is set to Read and <paramref name="capacity"/> is larger than the length of the file.</para>
+        /// </exception>
+        /// <exception cref="ArgumentNullException"><paramref name="fileHandle"/> is <see langword="null" />.</exception>
+        /// <exception cref="ArgumentOutOfRangeException">
+        /// <paramref name="capacity"/> is less than zero.
+        /// <para>-or-
+        /// <paramref name="capacity"/> is less than the file size.</para>
+        /// <para>-or-
+        /// <paramref name="access"/> is not a valid <see cref="MemoryMappedFileAccess"/> enumeration value.</para>
+        /// <para>-or-
+        /// <paramref name="inheritability"/> is not a valid <see cref="HandleInheritability"/> enumeration value.</para>
+        /// </exception>
```
The comment needs some minor tweaking:
* when referring to existing types we need to use `<see cref="$typeName"/>` everywhere, so IDE gives users the possibility to click the link and go to the type.
* same goes for enum values
* `<paramref name="capacity"/> is set to Write` was incorrect, I assume we need `<paramref name="access"/> is set to <see cref="MemoryMappedFileAccess.Write"/>`

BTW @carlossanlop @gewarren do we a doc that describes XML comments best practices? 

```suggestion
        /// <summary>
        /// Creates a memory-mapped file from an existing file using a <see cref="SafeFileHandle"/>,
        /// and the specified access mode, name, inheritability, and capacity.
        /// </summary>
        /// <param name="fileHandle">The <see cref="SafeFileHandle"/> to the existing file. Caller is
        /// responsible for disposing <paramref name="fileHandle"/> when <paramref name="leaveOpen"/> is <see langword="true" /> (otherwise,
        /// automatically disposed by the <see cref="MemoryMappedFile"/>). </param>
        /// <param name="mapName">A name to assign to the memory-mapped file, or <see langword="null" /> for a
        /// <see cref="MemoryMappedFile"/> that you do not intend to share across processes.</param>
        /// <param name="capacity">The maximum size, in bytes, to allocate to the memory-mapped file.
        /// Specify 0 to set the capacity to the size of the file.</param>
        /// <param name="access">One of the enumeration values that specifies the type of access allowed
        /// to the memory-mapped file.
        /// <para>This parameter can't be set to <see cref="MemoryMappedFileAccess.Write"/></para></param>
        /// <param name="inheritability">One of the enumeration values that specifies whether a handle
        /// to the memory-mapped file can be inherited by a child process. The default is <see cref="HandleInheritability.None"/>.</param>
        /// <param name="leaveOpen">A value that indicates whether to close the source file handle when
        /// the <see cref="MemoryMappedFile"/> is disposed.</param>
        /// <returns>A memory-mapped file that has the specified characteristics.</returns>
        /// <exception cref="ArgumentException">
        /// <para><paramref name="mapName"/> is <see langword="null" /> or an empty string.</para>
        /// <para>-or-</para>
        /// <para><paramref name="capacity"/> and the length of the file are zero.</para>
        /// <para>-or-</para>
        /// <para><paramref name="access"/> is set to <see cref="MemoryMappedFileAccess.Write"/>, which is not allowed.</para>
        /// <para>-or-</para>
        /// <para><paramref name="access"/> is set to <see cref="MemoryMappedFileAccess.Read"/> and <paramref name="capacity"/> is larger than the length of the file.</para>
        /// </exception>
        /// <exception cref="ArgumentNullException"><paramref name="fileHandle"/> is <see langword="null" />.</exception>
        /// <exception cref="ArgumentOutOfRangeException">
        /// <para><paramref name="capacity"/> is less than zero.</para>
        /// <para>-or-</para>
        /// <para><paramref name="capacity"/> is less than the file size.</para>
        /// <para>-or-</para>
        /// <para><paramref name="access"/> is not a valid <see cref="MemoryMappedFileAccess"/> enumeration value.</para>
        /// <para>-or-</para>
        /// <para><paramref name="inheritability"/> is not a valid <see cref="HandleInheritability"/> enumeration value.</para>
        /// </exception>
```

gewarren replied:
@adamsitnik We do have a guide, but I believe it's internal only: https://review.learn.microsoft.com/en-us/help/contribute-ref/how-to-write-net-docs?branch=main

adamsitnik replied:
@gewarren this guide is great, we should open source it!

gewarren replied:
@adamsitnik Just published the /// guide to the public contributor guide!
https://learn.microsoft.com/contribute/dotnet/api-documentation

adamsitnik commented on src/libraries/System.IO.MemoryMappedFiles/tests/MemoryMappedFile.CreateFromFile.Tests.cs:
```diff
@@ -36,7 +36,17 @@ public void InvalidArguments_Path()
+        [Fact]
+        public void InvalidArguments_SafeFileHandle()
+        {
+            // null is an invalid stream
```
nit: minor copy-paste error

```suggestion
            // null is an invalid handle
```

adamsitnik commented on src/libraries/System.IO.MemoryMappedFiles/tests/MemoryMappedFile.CreateFromFile.Tests.cs:
```diff
@@ -103,6 +113,19 @@ public void InvalidArguments_Access()
+
+            // Test the same things, but with a SafeFileHandle
+            using (TempFile file = new TempFile(GetTestFilePath()))
+            using (FileStream fs = File.Open(file.Path, FileMode.Open))
+            {
+                SafeFileHandle fileHandle = fs.SafeFileHandle;
```
This code is 100% correct, but we can improve it by not creating the `FileStream` and then using it's `SafeFileHandle`, but rather using the `File.OpenHandle` and work with handle directly.

The reason why I am asking for this change is that when we introduce new APIs we often don't provide docs with examples for them, so some users use our tests as examples. And if they want to use the new `SafeFileHandle` based APIs, they should not be creating `FileStream` up-front. `FileStream` is more expensive in terms of allocations (it's simply bigger and might allocate it's own buffer), moreover accessing `FileStream.SafeFileHandle` is performing a sys-call, which is expensive:

https://github.com/dotnet/runtime/blob/9eabb99670d4c0f26010451833b8f62ebbe60e32/src/libraries/System.Private.CoreLib/src/System/IO/Strategies/OSFileStreamStrategy.cs#L96-L105

```suggestion
            using (SafeFileHandle fileHandle = File.OpenHandle(file.Path, FileMode.Open))
            {
```

adamsitnik reviewed (commented):
The code looks very good, PTAL at my comments. Again thank you for your contribution @amalabey !

amalabey committed:
Fix xmldoc formatting issues

amalabey committed:
Fix copy paste error

amalabey committed:
Use fileHandle from File.OpenHandle

amalabey committed:
Merge remote-tracking branch 'upstream/main' into issue-79187

amalabey committed:
Merge branch 'issue-79187' of github.com:amalabey/runtime into issue-79187

adamsitnik reviewed (approved):
LGTM, thank you very much for your contribution @amalabey !
